src/persist/src/indexed/future.rs

```diff
@@ -149,9 +219,67 @@ impl<K: Data, V: Data> BlobFuture<K, V> {
             .into());
         }
         self.ts_lower = new_ts_lower;
-        // TODO: Actually delete the data.
+        self.evict();
         Ok(())
     }
+
+    /// Remove all batches containing only data strictly before the Future's time
+    /// lower bound.
+    fn evict(&mut self) {
+        // TODO: actually physically free the old batches.
+        let ts_lower = self.ts_lower.clone();
+        self.batches.retain(|b| ts_lower.less_equal(&b.ts_upper));
+    }
+
+    /// Create a new [BlobFutureBatch] from `batch` containing only the subset of
+    /// updates at or in advance of the Future's time lower bound.
+    ///
+    /// `batch` is assumed not be eligible for eviction at the time of this function
+    /// call, and to satisy all [BlobFutureBatch] invariants.
+    fn trim<L: Blob>(
+        &mut self,
+        batch: BlobFutureBatchMeta,
+        blob: &mut BlobCache<K, V, L>,
+    ) -> Result<BlobFutureBatchMeta, Error> {
+        // Sanity check that batch cannot be evicted
+        debug_assert!(self.ts_lower.less_equal(&batch.ts_upper));
+        let mut updates = vec![];
+
+        updates.extend(
+            blob.get_future_batch(&batch.key)?
+                .updates
+                .iter()
+                .filter(|(_, ts, _)| self.ts_lower.less_equal(ts))
+                .cloned(),
+        );
+        debug_assert!(!updates.is_empty());
+        let new_batch = BlobFutureBatch {
+            desc: batch.desc,
+            updates,
+        };
+
+        self.write_batch(new_batch, blob)
+    }
+
+    /// Take one step towards shrinking the representation of this future.
+    ///
+    /// Returns true if the trace was modified, false otherwise.
+    pub fn step<L: Blob>(&mut self, blob: &mut BlobCache<K, V, L>) -> Result<bool, Error> {
+        self.evict();
+
+        for (idx, batch) in self.batches.iter_mut().enumerate() {
+            // We can trim data out of the batch if it contains data at times < ts_lower.
+            if !self.ts_lower.less_equal(&batch.ts_lower) {
+                let batch = batch.clone();
+                let new_batch = self.trim(batch, blob)?;
+                self.batches.push(new_batch);
```

Comment:
if we don't need the old batch, can't we just assign the old one? that'd be more straightforward to follow

    self.batches[idx] = new_batch;

And if we do end up needing it, I think we can do a mem::swap